martian/src/metadata.rs

```diff
@@ -335,28 +343,27 @@ fn make_metadata_file_path(metadata_dir: &Path, name: &str) -> PathBuf {
     metadata_dir.join([METADATA_PREFIX, name].concat())
 }
 
-/// Manage shared access to a metadata file.
-#[derive(Debug, Clone)]
-pub(crate) struct SharedFile(Arc<Mutex<PathBuf>>);
-
-impl SharedFile {
-    pub fn new(path: PathBuf) -> Self {
-        Self(Arc::new(Mutex::new(path)))
+pub(crate) type SharedMetadata = Arc<Mutex<Metadata>>;
+
+/// Write a new file atomically, by first writing to a temp file, fsync, then rename.
+///
+/// The temp version will be the provided path with .tmp appended.
+fn fully_atomic_write(
+    path: &Path,
+    write: impl FnOnce(&mut dyn Write) -> std::io::Result<()>,
+) -> std::io::Result<()> {
+    let mut tmp_path: OsString = path.into();
+    tmp_path.push(".tmp");
+    let mut f = File::create(&tmp_path)?;
+    let temp_write_result = write(&mut f).and_then(|_| f.sync_all());
```

Comment:
This function should really just be taking in a `&[u8]` rather than a closure, because as is it's still possible for the caller to do something naughty (like for example printing an error message) during the closure.

Reply:
I cannot disagree more strongly.  Leaving aside the fact that the journal write should probably not be using this function in the first place, if can always ignore the error coming from this function - you don't need to inject a closure in here to deal with that.

Reply:
This allows us to ignore the failure to write what is probably (most of the time) unnecessary contents to the file, but *not* ignore an error coming from the rename call. One of those is probably not important, but the other one definitely is.